opencensus/trace/exporters/span_proto_exporter/span_proto_exporter.py

```diff
@@ -0,0 +1,196 @@
+import grpc
+import os
+import socket
+from threading import Lock
+from opencensus.trace.exporters import base
+from opencensus.trace.exporters.gen.opencensusd.agent.common.v1 import (
+    common_pb2
+)
+from opencensus.trace.exporters.gen.opencensusd.agent.trace.v1 import (
+    trace_service_pb2,
+    trace_service_pb2_grpc
+)
+from opencensus.trace.exporters.transports import sync
+from opencensus.trace.exporters.span_proto_exporter import utils
+
+# Default agent endpoint
+DEFAULT_ENDPOINT = 'localhost:50051'
+
+# OpenCensus Version
+VERSION = '0.1.6'
+
+
+class SpanProtoExporter(base.Exporter):
+    """Export the spans by sending them to opencensus agent.
+
+    :type service_name: str
+    :param service_name: name of the service
+
+    :type host_name: str
+    :param host_name: name of the host (machine or host name)
+
+    :type endpoint: str
+    :param endpoint: opencensus agent endpoint.
+
+    :type client: class:`~.trace_service_pb2_grpc.TraceServiceStub`
+    :param client: OpenCensusD client.
+
+    :type transport: :class:`type`
+    :param transport: Class for creating new transport objects. It should
+                      extend from the base :class:`.Transport` type and
+                      implement :meth:`.Transport.export`. Defaults to
+                      :class:`.SyncTransport`. The other option is
+                      :class:`.BackgroundThreadTransport`.
+    """
+
+    def __init__(
+            self,
+            service_name,
+            host_name=None,
+            endpoint=None,
+            client=None,
+            transport=sync.SyncTransport):
+        self.transport = transport(self)
+
+        self.endpoint = DEFAULT_ENDPOINT if endpoint is None else endpoint
+
+        if client is None:
+            self.channel = grpc.insecure_channel(self.endpoint)
+            self.client = trace_service_pb2_grpc.TraceServiceStub(
+                channel=self.channel)
+        else:
+            self.client = client
+
+        self.node = common_pb2.Node(
+            identifier=common_pb2.ProcessIdentifier(
+                host_name=socket.gethostname() if host_name is None
+                else host_name,
+                pid=os.getpid(),
+                start_timestamp=utils.proto_ts_from_datetime(
+                    datetime.datetime.now())
+            ),
+            library_info=common_pb2.LibraryInfo(
+                language=common_pb2.LibraryInfo.Language.Value('PYTHON'),
+                version=VERSION
+            ),
+            service_info=common_pb2.ServiceInfo(name=service_name))
+
+        self.send_node_in_config = True
+        self.send_node_in_export = True
+
+    def emit(self, span_datas):
+        """
+        :type span_datas: list of :class:
+            `~opencensus.trace.span_data.SpanData`
+        :param list of opencensus.trace.span_data.SpanData span_datas:
+            SpanData tuples to emit
+        """
+
+        try:
+            responses = self.client.Export(
```

Comment:
I think we want to keep the RPC (stream) alive for the entire time, not just the channel and stub. And the protocol says to send the node once per RPC stream not per channel, I need to investigate how to do this in python, but if we cannot easily do this we should always send the Node.

Reply:
My understanding was that until grpc is terminated because of error, it is kept alive. But I don't have enough grpc knowledge and can't find detailed documentation for python. I'd appreciate any help with it.

Let's send Node with every request for now and optimize later if possible.